### PR TITLE
Use a full ETHAssetHolder during pushOutcomeAndTransferAll test

### PR DESCRIPTION
**Pull request by geoknee**
Currently, we use a test contract derived from the base AssetHolder contract. This means that funds are not actually transferred out of the contract during testing, resulting in an underestimate of gas consumed. 

This PR rectifies that by using a contract derived from the full ETHAssetHolder.

## Changes [Optional] 
Because ETH actually gets transferred now, the test script must also deposit ETH into the ETHAssetHolders. (There are two deployed for our test suite).

Also changed is the
- python script
- generated png (and svg) used in our blog post

We may use the updated image to publish a new post updating the community on (dollar) gas costs in our system. 

Note: the (gas) gas costs have changed over recent months due to many small changes to the contracts and solc compiler options. 

---
### Code quality
- [x] I have written clear commit messages
- [x] I have performed a self-review of my own code
- [ ] This change does not have an unduly wide scope
- [x] I have separated logic changes from refactor changes (formatting, renames, etc.)
- [x] I have commented my code wherever necessary (can be 0)
- [x] I have added tests that prove my fix is effective or that my feature works, if necessary
### Project management
- [x] I have applied the [appropriate labels](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#b304e56734a74dfbb341b8b4b27b1c0c)
- [x] I have [linked to all relevant issues](https://help.zenhub.com/support/solutions/articles/43000010350-connecting-pull-requests-to-github-issues) (can be 0)
- [x] I have [linked to all dependent issues](https://help.zenhub.com/support/solutions/articles/43000010349-create-github-issue-dependencies) (can be 0)
- [x] I have assigned myself to this PR
- [x] I have chosen the appropriate [pipeline](https://www.notion.so/Team-working-agreements-2a95c926bb5642e5a5c42e4b74a9dd24#d534e68e7edc46fe8a4cda61b2258c4e) on zenhub for the linked issue

### CODE REVIEW

File: packages/nitro-protocol/contracts/test/TestEthAssetHolder2.sol [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.7.4;
+pragma experimental ABIEncoderV2;
+import '../ETHAssetHolder.sol';
+
+/**
+ * @dev This contract is a clone of the ETHAssetHolder contract. It will be initialized to point to the TestNitroAdjudicator.
+ */
+contract TestEthAssetHolder is ETHAssetHolder {
+    uint256 public dummy2; // this is simply to make the contract have distinct bytecode from the ETHAssetHolder (otherwise bytecode verification can fail)
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/contracts/test/TestEthAssetHolder2.sol:10]: The only difference between this file and `TestEthAssetHolder` seems to be `dummy` vs `dummy2`. What's the purpose of this file?

Comment (andrewgordstewart) [reply packages/nitro-protocol/contracts/test/TestEthAssetHolder2.sol:10]: Ok, I figured it out by reading the tests. I think there are tests that use multiple EthAssetHolders, and you wanted to use different addresses for them? It's hard to tell for sure.

File: packages/nitro-protocol/deployment/deploy.ts [hunk 2]
```diff
@@ -56,6 +57,11 @@ export async function deploy(): Promise<Record<string, string>> {
     {},
     TEST_NITRO_ADJUDICATOR_ADDRESS
   );
+  const TEST_ETH_ASSET_HOLDER2_ADDRESS = await deployer.deploy(
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/deployment/deploy.ts:60]: Is there a risk of `deploy.js` being used in "production"? (to deploy to mainnet, lets say?)

The `TestEthAssetHolder` contracts now have `setHoldings`, which mean we only want to deploy it to a test environment. The reason for asking then, is to determine: how can we ensure that `TestEthAssetHolder` is only used in a test environment?

File: packages/nitro-protocol/contracts/test/TestEthAssetHolder.sol [hunk 3]
```diff
@@ -2,12 +2,22 @@
...
     uint256 public dummy; // this is simply to make the contract have distinct bytecode from the ETHAssetHolder (otherwise bytecode verification can fail)
 
+    /**
+     * @dev Manually set the holdings mapping to a given amount for a given channelId.  Shortcuts the deposit workflow (ONLY USE IN A TESTING ENVIRONMENT)
+     * @param channelId Unique identifier for a state channel.
+     * @param amount The number of assets that should now be "escrowed: against channelId
+     */
+    function setHoldings(bytes32 channelId, uint256 amount) external {
+        holdings[channelId] = amount;
+    }
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/contracts/test/TestEthAssetHolder.sol:19]: I don't really understand why this change makes it a "full eth asset holder".

File: packages/nitro-protocol/test/contracts/NitroAdjudicator/concludePushOutcomeAndTransferAll.test.ts [hunk 4]
```diff
@@ -173,6 +173,22 @@ describe('concludePushOutcomeAndTransferAll', () => {
...
+        ).wait();
+      }
+      if ('ETH2' in heldBefore) {
+        // Transfer some funds in to the other ETHAssetHolder. Use zero channel address for ease.
+        await (
+          await EthAssetHolder2.deposit(channelId, '0x00', heldBefore.ETH2.c, {
+            value: heldBefore.ETH2.c,
+          })
+        ).wait();
+      }
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/test/contracts/NitroAdjudicator/concludePushOutcomeAndTransferAll.test.ts]: Is this the change that makes means we are now "[using] a full eth asset holder in tests"?

File: packages/nitro-protocol/test/contracts/NitroAdjudicator/concludePushOutcomeAndTransferAll.test.ts [hunk 5]
```diff
@@ -1,8 +1,8 @@
 import {expectRevert} from '@statechannels/devtools';
-import {Contract, Wallet, ethers, BigNumber} from 'ethers';
+import {Contract, Wallet, ethers, BigNumber, utils, constants} from 'ethers';
 
-import AssetHolderArtifact1 from '../../../artifacts/contracts/test/TESTAssetHolder.sol/TESTAssetHolder.json';
-import AssetHolderArtifact2 from '../../../artifacts/contracts/test/TESTAssetHolder2.sol/TESTAssetHolder2.json';
+import EthAssetHolderArtifact1 from '../../../artifacts/contracts/test/TestEthAssetHolder.sol/TestEthAssetHolder.json';
+import EthAssetHolderArtifact2 from '../../../artifacts/contracts/test/TestEthAssetHolder2.sol/TestEthAssetHolder.json';
```
Comment (andrewgordstewart) [inline packages/nitro-protocol/test/contracts/NitroAdjudicator/concludePushOutcomeAndTransferAll.test.ts:5]: Ok, now I think I understand. The diff made me think there was a rename happening, combined with other changes. (There is a rename, as `AssetHolder1` is renamed to `EthAssetHolder1`.)

But this change is _not_ a rename, and you are importing

8b173676315f947ae8f9730b5365ea03526dcaa3 would have been more clear if it were written something like

```
use TestEthAssetHolder in pushOutcomeAndTransferAll test

The TESTAssetHolder that is currently used is not a real asset holder,
since it doesn't send funds to the destination. This is because it
extends the base AssetHolder contract.

We switch to TestEthAssetHolder & TestEthAssetHolder2 to
get a more accurate measure of gas consumption.
```

**Review by andrewgordstewart — APPROVED** [review]
Assuming that you're going to use squash-and-merge, I might suggest clarifying the resulting title & commit message with something like

```
use TestEthAssetHolder in pushOutcomeAndTransferAll test

The TESTAssetHolder that is currently used is not a real asset holder,
since it doesn't send funds to the destination. This is because it
extends the base AssetHolder contract.

We switch to TestEthAssetHolder & TestEthAssetHolder2, which
extend the EthAssetHolder contract. This gives a more accurate
measure of gas consumption.
```

File: packages/nitro-protocol/deployment/deploy.ts (hunk 2 above)
Comment (geoknee) [reply packages/nitro-protocol/deployment/deploy.ts:60]: I'm going to address this in a separate PR.

File: packages/nitro-protocol/contracts/test/TestEthAssetHolder.sol (hunk 3 above)
Comment (geoknee) [reply packages/nitro-protocol/contracts/test/TestEthAssetHolder.sol:19]: This isn't. This is just a change that was necessary to use the full eth asset holder in the current test. I have another PR in the pipe that will actually do away with this contract altogether.

File: packages/nitro-protocol/test/contracts/NitroAdjudicator/concludePushOutcomeAndTransferAll.test.ts (hunk 4 above)
Comment (geoknee) [reply packages/nitro-protocol/test/contracts/NitroAdjudicator/concludePushOutcomeAndTransferAll.test.ts]: No, this is just another follow on change that is necessary, given that we now use the full eth asset holder in tests. You can't pay ETH out, unless you have paid some in :-)
